File: android/src/main/java/com/openfoodfacts/OpenFoodFactsModule.kt
package com.openfoodfacts

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.module.annotations.ReactModule
import com.facebook.react.turbomodule.core.interfaces.TurboModule
import com.openfoodfacts.api.OpenFoodFactsApiClient
import com.openfoodfacts.api.OpenFoodFactsConfigStore
import com.openfoodfacts.api.OpenFoodFactsException
import com.openfoodfacts.api.OpenFoodFactsMappers
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

@ReactModule(name = OpenFoodFactsModule.NAME)
class OpenFoodFactsModule(reactContext: ReactApplicationContext) :
  NativeOpenFoodFactsSpec(reactContext), TurboModule {
  private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
  private val configStore = OpenFoodFactsConfigStore.create(reactContext)
  private val apiClient = OpenFoodFactsApiClient(configStore = configStore)

  override fun getName(): String = NAME

  override fun getRuntimeInfo(promise: Promise) {
    val runtimeInfo = Arguments.createMap().apply {
      putString("platform", "android")
      putString("moduleName", NAME)
      putBoolean("newArchitecture", true)
      putString("nativeVersion", NATIVE_VERSION)
    }

    promise.resolve(runtimeInfo)
  }

  override fun configure(config: ReadableMap, promise: Promise) {
    runPromise(promise) {
      OpenFoodFactsMappers.toWritableMap(apiClient.configure(OpenFoodFactsMappers.parseConfigPatch(config)))
    }
  }

  override fun getConfig(promise: Promise) {
    runPromise(promise) {
      OpenFoodFactsMappers.toWritableMap(apiClient.getConfig())
    }
  }

  override fun resetConfig(promise: Promise) {
    runPromise(promise) {
      OpenFoodFactsMappers.toWritableMap(apiClient.resetConfig())
    }
  }

  override fun getProduct(query: ReadableMap, promise: Promise) {
    runPromise(promise) {
      OpenFoodFactsMappers.toWritableMap(apiClient.getProduct(OpenFoodFactsMappers.parseProductQuery(query)))
    }
  }

  override fun getOrderedNutrients(promise: Promise) {
    runPromise(promise) {
      OpenFoodFactsMappers.toWritableOrderedNutrients(apiClient.getOrderedNutrients())
    }
  }

  override fun getNutrientMetadata(promise: Promise) {
    runPromise(promise) {
      OpenFoodFactsMappers.toWritableMap(apiClient.getNutrientMetadata())
    }
  }

  override fun getSuggestions(query: ReadableMap, promise: Promise) {
    runPromise(promise) {
      OpenFoodFactsMappers.toWritableStringArray(apiClient.getSuggestions(OpenFoodFactsMappers.parseSuggestionsQuery(query)))
    }
  }

  override fun saveProduct(request: ReadableMap, promise: Promise) {
    runPromise(promise) {
      OpenFoodFactsMappers.toWritableMap(apiClient.saveProduct(OpenFoodFactsMappers.parseSaveProductRequest(request)))
    }
  }

  override fun patchProduct(request: ReadableMap, promise: Promise) {
    runPromise(promise) {
      apiClient.patchProduct(OpenFoodFactsMappers.parseProductPatchRequest(request))
    }
  }

  override fun uploadProductImage(request: ReadableMap, promise: Promise) {
    runPromise(promise) {
      apiClient.uploadProductImage(OpenFoodFactsMappers.parseProductImageUploadRequest(request))
    }
  }

  override fun deleteProductImage(request: ReadableMap, promise: Promise) {
    runPromise(promise) {
      apiClient.deleteProductImage(OpenFoodFactsMappers.parseProductImageDeleteRequest(request))
    }
  }

  override fun canonicalizeTags(query: ReadableMap, promise: Promise) {
    runPromise(promise) {
      OpenFoodFactsMappers.toWritableMap(apiClient.canonicalizeTags(OpenFoodFactsMappers.parseTaxonomyCanonicalizeQuery(query)))
    }
  }

  override fun displayTags(query: ReadableMap, promise: Promise) {
    runPromise(promise) {
      OpenFoodFactsMappers.toWritableMap(apiClient.displayTags(OpenFoodFactsMappers.parseTaxonomyDisplayQuery(query)))
    }
  }

  override fun getTagKnowledge(query: ReadableMap, promise: Promise) {
    runPromise(promise) {
      apiClient.getTagKnowledge(OpenFoodFactsMappers.parseTagKnowledgeQuery(query))
    }
  }

  override fun revertProduct(request: ReadableMap, promise: Promise) {
    runPromise(promise) {
      apiClient.revertProduct(OpenFoodFactsMappers.parseProductRevertRequest(request))
    }
  }

  override fun getExternalSources(promise: Promise) {
    runPromise(promise) {
      apiClient.getExternalSources()
    }
  }

  override fun getPreferences(promise: Promise) {
    runPromise(promise) {
      apiClient.getPreferences()
    }
  }

  override fun getAttributeGroups(promise: Promise) {
    runPromise(promise) {
      apiClient.getAttributeGroups()
    }
  }

  override fun invalidate() {
    scope.cancel()
    super.invalidate()
  }

  private fun runPromise(promise: Promise, block: suspend () -> Any) {
    scope.launch {
      try {
        promise.resolve(block())
      } catch (error: OpenFoodFactsException) {
        promise.reject(error.errorCode, error.message ?: "Open Food Facts request failed.", error)
      } catch (error: Exception) {
        promise.reject("E_NETWORK", error.message ?: "Unexpected Open Food Facts failure.", error)
      }
    }
  }

  companion object {
    const val NAME = "OpenFoodFacts"
    private const val NATIVE_VERSION = "0.1.0"
  }
}

File: ios/OpenFoodFactsModule.swift
import Foundation
import React

@objc(OpenFoodFactsModule)
public final class OpenFoodFactsModule: NSObject {
  private let client = OpenFoodFactsIOSClient()

  @objc
  public static func requiresMainQueueSetup() -> Bool {
    false
  }

  @objc(getRuntimeInfo:reject:)
  public func getRuntimeInfo(
    _ resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    resolve([
      "platform": "ios",
      "moduleName": "OpenFoodFacts",
      "newArchitecture": true,
      "nativeVersion": "0.1.0",
    ])
  }

  @objc(configure:resolve:reject:)
  public func configure(
    _ config: NSDictionary,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    Task {
      do {
        resolve(try await client.configure(asSwiftDictionary(config)))
      } catch let error as OpenFoodFactsModuleError {
        reject(error.code, error.message, error.asNSError())
      } catch {
        reject("E_NETWORK", error.localizedDescription, error)
      }
    }
  }

  @objc(getConfig:reject:)
  public func getConfig(
    _ resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    Task {
      resolve(await client.getConfig())
    }
  }

  @objc(resetConfig:reject:)
  public func resetConfig(
    _ resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    Task {
      resolve(await client.resetConfig())
    }
  }

  @objc(getProduct:resolve:reject:)
  public func getProduct(
    _ query: NSDictionary,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    Task {
      do {
        resolve(try await client.getProduct(asSwiftDictionary(query)))
      } catch let error as OpenFoodFactsModuleError {
        reject(error.code, error.message, error.asNSError())
      } catch {
        reject("E_NETWORK", error.localizedDescription, error)
      }
    }
  }

  @objc(getOrderedNutrients:reject:)
  public func getOrderedNutrients(
    _ resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    Task {
      do {
        resolve(try await client.getOrderedNutrients())
      } catch let error as OpenFoodFactsModuleError {
        reject(error.code, error.message, error.asNSError())
      } catch {
        reject("E_NETWORK", error.localizedDescription, error)
      }
    }
  }

  @objc(getNutrientMetadata:reject:)
  public func getNutrientMetadata(
    _ resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    Task {
      do {
        resolve(try await client.getNutrientMetadata())
      } catch let error as OpenFoodFactsModuleError {
        reject(error.code, error.message, error.asNSError())
      } catch {
        reject("E_NETWORK", error.localizedDescription, error)
      }
    }
  }

  @objc(getSuggestions:resolve:reject:)
  public func getSuggestions(
    _ query: NSDictionary,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    Task {
      do {
        resolve(try await client.getSuggestions(asSwiftDictionary(query)))
      } catch let error as OpenFoodFactsModuleError {
        reject(error.code, error.message, error.asNSError())
      } catch {
        reject("E_NETWORK", error.localizedDescription, error)
      }
    }
  }

  @objc(search:resolve:reject:)
  public func search(
    _ query: NSDictionary,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    Task {
      do {
        resolve(try await client.search(asSwiftDictionary(query)))
      } catch let error as OpenFoodFactsModuleError {
        reject(error.code, error.message, error.asNSError())
      } catch {
        reject("E_NETWORK", error.localizedDescription, error)
      }
    }
  }

  @objc(saveProduct:resolve:reject:)
  public func saveProduct(
    _ request: NSDictionary,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    Task {
      do {
        resolve(try await client.saveProduct(asSwiftDictionary(request)))
      } catch let error as OpenFoodFactsModuleError {
        reject(error.code, error.message, error.asNSError())
      } catch {
        reject("E_NETWORK", error.localizedDescription, error)
      }
    }
  }

  @objc(patchProduct:resolve:reject:)
  public func patchProduct(
    _ request: NSDictionary,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    Task {
      do {
        resolve(try await client.patchProduct(asSwiftDictionary(request)))
      } catch let error as OpenFoodFactsModuleError {
        reject(error.code, error.message, error.asNSError())
      } catch {
        reject("E_NETWORK", error.localizedDescription, error)
      }
    }
  }

  @objc(uploadProductImage:resolve:reject:)
  public func uploadProductImage(
    _ request: NSDictionary,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    Task {
      do {
        resolve(try await client.uploadProductImage(asSwiftDictionary(request)))
      } catch let error as OpenFoodFactsModuleError {
        reject(error.code, error.message, error.asNSError())
      } catch {
        reject("E_NETWORK", error.localizedDescription, error)
      }
    }
  }

  @objc(deleteProductImage:resolve:reject:)
  public func deleteProductImage(
    _ request: NSDictionary,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    Task {
      do {
        resolve(try await client.deleteProductImage(asSwiftDictionary(request)))
      } catch let error as OpenFoodFactsModuleError {
        reject(error.code, error.message, error.asNSError())
      } catch {
        reject("E_NETWORK", error.localizedDescription, error)
      }
    }
  }

  @objc(canonicalizeTags:resolve:reject:)
  public func canonicalizeTags(
    _ query: NSDictionary,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    Task {
      do {
        resolve(try await client.canonicalizeTags(asSwiftDictionary(query)))
      } catch let error as OpenFoodFactsModuleError {
        reject(error.code, error.message, error.asNSError())
      } catch {
        reject("E_NETWORK", error.localizedDescription, error)
      }
    }
  }

  @objc(displayTags:resolve:reject:)
  public func displayTags(
    _ query: NSDictionary,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    Task {
      do {
        resolve(try await client.displayTags(asSwiftDictionary(query)))
      } catch let error as OpenFoodFactsModuleError {
        reject(error.code, error.message, error.asNSError())
      } catch {
        reject("E_NETWORK", error.localizedDescription, error)
      }
    }
  }

  @objc(getTagKnowledge:resolve:reject:)
  public func getTagKnowledge(
    _ query: NSDictionary,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    Task {
      do {
        resolve(try await client.getTagKnowledge(asSwiftDictionary(query)))
      } catch let error as OpenFoodFactsModuleError {
        reject(error.code, error.message, error.asNSError())
      } catch {
        reject("E_NETWORK", error.localizedDescription, error)
      }
    }
  }

  @objc(revertProduct:resolve:reject:)
  public func revertProduct(
    _ request: NSDictionary,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    Task {
      do {
        resolve(try await client.revertProduct(asSwiftDictionary(request)))
      } catch let error as OpenFoodFactsModuleError {
        reject(error.code, error.message, error.asNSError())
      } catch {
        reject("E_NETWORK", error.localizedDescription, error)
      }
    }
  }

  @objc(getExternalSources:reject:)
  public func getExternalSources(
    _ resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    Task {
      do {
        resolve(try await client.getExternalSources())
      } catch let error as OpenFoodFactsModuleError {
        reject(error.code, error.message, error.asNSError())
      } catch {
        reject("E_NETWORK", error.localizedDescription, error)
      }
    }
  }

  @objc(getPreferences:reject:)
  public func getPreferences(
    _ resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    Task {
      do {
        resolve(try await client.getPreferences())
      } catch let error as OpenFoodFactsModuleError {
        reject(error.code, error.message, error.asNSError())
      } catch {
        reject("E_NETWORK", error.localizedDescription, error)
      }
    }
  }

  @objc(getAttributeGroups:reject:)
  public func getAttributeGroups(
    _ resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    Task {
      do {
        resolve(try await client.getAttributeGroups())
      } catch let error as OpenFoodFactsModuleError {
        reject(error.code, error.message, error.asNSError())
      } catch {
        reject("E_NETWORK", error.localizedDescription, error)
      }
    }
  }
}

private func asSwiftDictionary(_ dictionary: NSDictionary) -> [String: Any] {
  dictionary as? [String: Any] ?? [:]
}
